replica/table: check memtable before discarding tombstone during read

lkshminarayanan opened the pull request:
On the read path, the compacting reader is applied only to the sstable
reader. This can cause an expired tombstone from an sstable to be purged
from the request before it has a chance to merge with deleted data in
the memtable leading to data resurrection.

Fix this by checking the memtables before deciding to purge tombstones
from the request on the read path. A tombstone will not be purged if a
key exists in any of the table's memtables with a minimum live timestamp
that is lower than the maximum purgeable timestamp.

Fixes #20916

`perf-simple-query` stats before and after this fix :

`build/Dev/scylla perf-simple-query --smp=1 --flush` :
```
// Before this Fix
// ---------------
94941.79 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59393 insns/op,   24029 cycles/op,        0 errors)
97551.14 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59376 insns/op,   23966 cycles/op,        0 errors)
96599.92 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59367 insns/op,   23998 cycles/op,        0 errors)
97774.91 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59370 insns/op,   23968 cycles/op,        0 errors)
97796.13 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59368 insns/op,   23947 cycles/op,        0 errors)

         throughput: mean=96932.78 standard-deviation=1215.71 median=97551.14 median-absolute-deviation=842.13 maximum=97796.13 minimum=94941.79
instructions_per_op: mean=59374.78 standard-deviation=10.78 median=59369.59 median-absolute-deviation=6.36 maximum=59393.12 minimum=59367.02
  cpu_cycles_per_op: mean=23981.67 standard-deviation=32.29 median=23967.76 median-absolute-deviation=16.33 maximum=24029.38 minimum=23947.19

// After this Fix
// --------------
95313.53 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59392 insns/op,   24058 cycles/op,        0 errors)
97311.48 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59375 insns/op,   24005 cycles/op,        0 errors)
98043.10 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59381 insns/op,   23941 cycles/op,        0 errors)
96750.31 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59396 insns/op,   24025 cycles/op,        0 errors)
93381.21 tps ( 71.1 allocs/op,   0.0 logallocs/op,  14.1 tasks/op,   59390 insns/op,   24097 cycles/op,        0 errors)

         throughput: mean=96159.93 standard-deviation=1847.88 median=96750.31 median-absolute-deviation=1151.55 maximum=98043.10 minimum=93381.21
instructions_per_op: mean=59386.60 standard-deviation=8.78 median=59389.55 median-absolute-deviation=6.02 maximum=59396.40 minimum=59374.73
  cpu_cycles_per_op: mean=24025.13 standard-deviation=58.39 median=24025.17 median-absolute-deviation=32.67 maximum=24096.66 minimum=23941.22
```

This PR fixes a regression introduced in ce96b472d3 and should be backported to older versions.

raphaelsc commented on replica/table.cc:
```diff
@@ -2517,6 +2517,24 @@ table::make_partition_presence_checker(lw_shared_ptr<const sstables::sstable_set
+max_purgeable_fn table::get_max_purgeable_fn_for_compacting_reader() {
+    return [this] (const dht::decorated_key& dk, ::is_shadowable is_shadowable) {
+        auto max_purgeable_timestamp = api::max_timestamp;
+
```
we can easily grab the storage group that overlaps with dk.token(). see storage_group_for_token(), so we reduce the search space.

raphaelsc commented on replica/table.cc:
```diff
@@ -2517,6 +2517,24 @@ table::make_partition_presence_checker(lw_shared_ptr<const sstables::sstable_set
+        });
+
+        return max_purgeable_timestamp;
+    };
```
could you please measure perf impact after this change using perf-simple-query (there's an option to flush memtable and bypass cache)?

lkshminarayanan replied:
`/build/Dev/scylla perf-simple-query --flush` :
```
// Before this Fix
// ---------------
761137.87 tps ( 74.2 allocs/op,   0.0 logallocs/op,  18.5 tasks/op,   46965 insns/op,   35265 cycles/op,        0 errors)
766813.17 tps ( 74.2 allocs/op,   0.0 logallocs/op,  18.4 tasks/op,   45922 insns/op,   34680 cycles/op,        0 errors)
1005809.42 tps ( 74.2 allocs/op,   0.0 logallocs/op,  18.4 tasks/op,   47534 insns/op,   35314 cycles/op,        0 errors)
1179927.13 tps ( 74.1 allocs/op,   0.0 logallocs/op,  18.3 tasks/op,   44717 insns/op,   33781 cycles/op,        0 errors)
1170860.88 tps ( 74.1 allocs/op,   0.0 logallocs/op,  18.3 tasks/op,   45694 insns/op,   34419 cycles/op,        0 errors)

         throughput: mean=976909.69 standard-deviation=206377.30 median=1005809.42 median-absolute-deviation=203017.44 maximum=1179927.13 minimum=761137.87
instructions_per_op: mean=46166.49 standard-deviation=1105.91 median=45921.69 median-absolute-deviation=798.90 maximum=47534.01 minimum=44717.04
  cpu_cycles_per_op: mean=34691.81 standard-deviation=636.37 median=34680.01 median-absolute-deviation=573.13 maximum=35313.90 minimum=33780.72

// After this Fix
// --------------
1229741.48 tps ( 74.1 allocs/op,   0.0 logallocs/op,  18.3 tasks/op,   44284 insns/op,   33345 cycles/op,        0 errors)
1168447.23 tps ( 74.2 allocs/op,   0.0 logallocs/op,  18.5 tasks/op,   46583 insns/op,   34777 cycles/op,        0 errors)
1164043.08 tps ( 74.2 allocs/op,   0.0 logallocs/op,  18.5 tasks/op,   46833 insns/op,   34831 cycles/op,        0 errors)
1161622.12 tps ( 74.2 allocs/op,   0.0 logallocs/op,  18.4 tasks/op,   46835 insns/op,   35396 cycles/op,        0 errors)
1142232.60 tps ( 74.2 allocs/op,   0.0 logallocs/op,  18.5 tasks/op,   48054 insns/op,   36021 cycles/op,        0 errors)

         throughput: mean=1173217.30 standard-deviation=33152.69 median=1164043.08 median-absolute-deviation=11595.18 maximum=1229741.48 minimum=1142232.60
instructions_per_op: mean=46517.74 standard-deviation=1374.30 median=46833.29 median-absolute-deviation=317.07 maximum=48053.84 minimum=44283.71
  cpu_cycles_per_op: mean=34874.14 standard-deviation=992.16 median=34830.70 median-absolute-deviation=521.90 maximum=36021.35 minimum=33345.14
```

lkshminarayanan replied:
With cache bypass :

`./build/Dev/scylla perf-simple-query --flush --bypass-cache` :

```
// Before this Fix
// ---------------
117014.45 tps (148.0 allocs/op, 506.0 logallocs/op,  46.9 tasks/op,  622579 insns/op,  329591 cycles/op,        0 errors)
119100.62 tps (147.6 allocs/op, 506.1 logallocs/op,  46.8 tasks/op,  615611 insns/op,  325732 cycles/op,        0 errors)
106061.77 tps (147.6 allocs/op, 506.0 logallocs/op,  46.7 tasks/op,  607200 insns/op,  323245 cycles/op,        0 errors)
54462.23 tps (147.8 allocs/op, 506.1 logallocs/op,  46.9 tasks/op,  618228 insns/op,  320103 cycles/op,        0 errors)
54918.77 tps (147.8 allocs/op, 506.0 logallocs/op,  46.9 tasks/op,  618525 insns/op,  320329 cycles/op,        0 errors)

         throughput: mean=90311.57 standard-deviation=32892.81 median=106061.77 median-absolute-deviation=28789.05 maximum=119100.62 minimum=54462.23
instructions_per_op: mean=616428.69 standard-deviation=5729.20 median=618228.19 median-absolute-deviation=2096.74 maximum=622578.53 minimum=607199.82
  cpu_cycles_per_op: mean=323800.26 standard-deviation=3977.95 median=323245.34 median-absolute-deviation=3471.14 maximum=329591.43 minimum=320103.09

// After this Fix
// --------------
78445.97 tps (148.3 allocs/op, 506.1 logallocs/op,  47.1 tasks/op,  624587 insns/op,  341380 cycles/op,        0 errors)
116746.21 tps (147.6 allocs/op, 506.2 logallocs/op,  46.8 tasks/op,  618510 insns/op,  329745 cycles/op,        0 errors)
115050.84 tps (147.6 allocs/op, 506.0 logallocs/op,  46.8 tasks/op,  623092 insns/op,  332330 cycles/op,        0 errors)
115856.59 tps (147.7 allocs/op, 506.0 logallocs/op,  46.8 tasks/op,  626256 insns/op,  334032 cycles/op,        0 errors)
113798.59 tps (147.7 allocs/op, 506.1 logallocs/op,  46.8 tasks/op,  622230 insns/op,  333158 cycles/op,        0 errors)

         throughput: mean=107979.64 standard-deviation=16545.38 median=115050.84 median-absolute-deviation=7876.95 maximum=116746.21 minimum=78445.97
instructions_per_op: mean=622934.82 standard-deviation=2909.37 median=623092.25 median-absolute-deviation=1651.73 maximum=626255.69 minimum=618509.96
  cpu_cycles_per_op: mean=334128.97 standard-deviation=4358.43 median=333158.19 median-absolute-deviation=1798.57 maximum=341379.85 minimum=329744.72
```

lkshminarayanan replied:
@raphaelsc The results are pasted above.

raphaelsc replied:
setting cpu frequency to a fixed rate might help with variance in results:

sudo cpupower frequency-set -g userspace
sudo cpupower frequency-set -f 2200MHz


do you think the regression comes from mt.contains_partition(dk)? for example, inst/op regresses ~1%. maybe it's that results are distubed, please rerun with that fixed frequency. if there's a regression and it comes mostly from that lookup, we can optimize common case by tracking max_sstable_timestamp (in every compaction group) and avoid the lookup for common scenarios where min memtable timestamp is higher than max sstable timestamps. perf record / report might reveal it.

raphaelsc replied:
I tend to think it's not the lookup though since memtable was flushed in this test and should be empty.

raphaelsc replied:
please include the results in the cover letter.

bhalevy replied:
@lkshminarayanan did you set the cpu frequency to the lowest value?
With performance / on-demand cpufreq governor you may see thermal effects as the cpu heats up and auto-tunes itself dynamically to a lower frequency, causing flakiness.

lkshminarayanan replied:
@bhalevy I took another set of readings with the cpu frequency set to the lowest value. I have pasted that into the PR cover letter. cc @raphaelsc

bhalevy replied:
Thanks it looks like there is a .5% degredation in read throughput.
Is this a trend or just jitter?
Is it expected?

raphaelsc replied:
"so I'm not sure why the PR causes a degradation when cache is not used." the patch will have no effect with bypass cache, since compacting reader is not involved at all. so I think this is about the results being disturbed by external factors.

raphaelsc replied:
@lkshminarayanan could you please modify perf_simple_query to write some keys to memtable after the last flush (see create_partitions)? so we can stress that lookup in the perf test. I think we might have to track max sstable timestamp for each group, to short circuit that lookup in common scenarios where memtable's data is newer than data in sstables.

lkshminarayanan replied:
okay

lkshminarayanan replied:
@raphaelsc I was finally able to modify `perf_simple_query` to test this PR as per your suggestion. 

The `perf_simple_query` now 
 - sets gc_grace_seconds = 0s to the schema
 - writes 10K partitions, with 33% tombstones
 - waits for a second to let tombstones expire
 - flushes all the tables and the commitlog
 - drops the sstable caches
 - writes 1K partitions in the memtable
 - and then finally runs the read test.

This is the result of `build/Dev/scylla perf-simple-query --smp=1 --flush --duration=300` :
```
// Before this Fix
// ---------------
42081.34 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.1 tasks/op,   70067 insns/op,   32260 cycles/op,        0 errors)
56577.90 tps ( 69.9 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   58029 insns/op,   24112 cycles/op,        0 errors)
58074.16 tps ( 69.8 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57945 insns/op,   23777 cycles/op,        0 errors)
58128.32 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57887 insns/op,   23775 cycles/op,        0 errors)
58102.11 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57919 insns/op,   23766 cycles/op,        0 errors)

...
...
         throughput: mean=58113.29 standard-deviation=948.09 median=58207.25 median-absolute-deviation=118.37 maximum=58460.52 minimum=42081.34
instructions_per_op: mean=57968.85 standard-deviation=701.93 median=57927.94 median-absolute-deviation=44.95 maximum=70066.91 minimum=57815.54
  cpu_cycles_per_op: mean=23770.54 standard-deviation=494.48 median=23738.03 median-absolute-deviation=42.94 maximum=32259.76 minimum=23646.28


// After this Fix
// --------------
41921.32 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   70069 insns/op,   32395 cycles/op,        0 errors)
57360.31 tps ( 69.8 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   58091 insns/op,   24066 cycles/op,        0 errors)
58126.02 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57907 insns/op,   23779 cycles/op,        0 errors)
58348.81 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57924 insns/op,   23676 cycles/op,        0 errors)
58266.16 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57898 insns/op,   23726 cycles/op,        0 errors)
...
...
         throughput: mean=58209.50 standard-deviation=964.09 median=58292.08 median-absolute-deviation=123.26 maximum=58661.90 minimum=41921.32
instructions_per_op: mean=57977.62 standard-deviation=701.51 median=57936.37 median-absolute-deviation=43.33 maximum=70068.51 minimum=57839.51
  cpu_cycles_per_op: mean=23739.14 standard-deviation=505.08 median=23709.30 median-absolute-deviation=42.39 maximum=32395.37 minimum=23573.21
```

WDYT? Do we have to  track max sstable timestamp for each group reduce to short circuit the memtable lookup ?

mykaul replied:
> The `perf_simple_query` now
> 
> * sets gc_grace_seconds = 0s to the schema
> * writes 10K partitions, with 33% tombstones
> * waits for a second to let tombstones expire
> * flushes all the tables and the commitlog
> * drops the sstable caches
> * writes 1K partitions in the memtable
> * and then finally runs the read test.

This is great, but it's not 'simple' anymore ;-)

How can we morph this into a now-not-so-simple perf tests?

raphaelsc replied:
> @raphaelsc I was finally able to modify `perf_simple_query` to test this PR as per your suggestion.
> 
> The `perf_simple_query` now
> 
> * sets gc_grace_seconds = 0s to the schema
> * writes 10K partitions, with 33% tombstones
> * waits for a second to let tombstones expire
> * flushes all the tables and the commitlog
> * drops the sstable caches
> * writes 1K partitions in the memtable
> * and then finally runs the read test.
> 
> This is the result of `build/Dev/scylla perf-simple-query --smp=1 --flush --duration=300` :
> 
> ```
> // Before this Fix
> // ---------------
> 42081.34 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.1 tasks/op,   70067 insns/op,   32260 cycles/op,        0 errors)
> 56577.90 tps ( 69.9 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   58029 insns/op,   24112 cycles/op,        0 errors)
> 58074.16 tps ( 69.8 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57945 insns/op,   23777 cycles/op,        0 errors)
> 58128.32 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57887 insns/op,   23775 cycles/op,        0 errors)
> 58102.11 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57919 insns/op,   23766 cycles/op,        0 errors)
> 
> ...
> ...
>          throughput: mean=58113.29 standard-deviation=948.09 median=58207.25 median-absolute-deviation=118.37 maximum=58460.52 minimum=42081.34
> instructions_per_op: mean=57968.85 standard-deviation=701.93 median=57927.94 median-absolute-deviation=44.95 maximum=70066.91 minimum=57815.54
>   cpu_cycles_per_op: mean=23770.54 standard-deviation=494.48 median=23738.03 median-absolute-deviation=42.94 maximum=32259.76 minimum=23646.28
> 
> 
> // After this Fix
> // --------------
> 41921.32 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   70069 insns/op,   32395 cycles/op,        0 errors)
> 57360.31 tps ( 69.8 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   58091 insns/op,   24066 cycles/op,        0 errors)
> 58126.02 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57907 insns/op,   23779 cycles/op,        0 errors)
> 58348.81 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57924 insns/op,   23676 cycles/op,        0 errors)
> 58266.16 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57898 insns/op,   23726 cycles/op,        0 errors)
> ...
> ...
>          throughput: mean=58209.50 standard-deviation=964.09 median=58292.08 median-absolute-deviation=123.26 maximum=58661.90 minimum=41921.32
> instructions_per_op: mean=57977.62 standard-deviation=701.51 median=57936.37 median-absolute-deviation=43.33 maximum=70068.51 minimum=57839.51
>   cpu_cycles_per_op: mean=23739.14 standard-deviation=505.08 median=23709.30 median-absolute-deviation=42.39 maximum=32395.37 minimum=23573.21
> ```
> 
> WDYT? Do we have to track max sstable timestamp for each group reduce to short circuit the memtable lookup ?

just to make sure, we now can see the calls to new function to take into account memtable timestamp, right?

raphaelsc replied:
> @mykaul My modifications are more of a hack needed to test this particular PR. We could probably add options to include tombstones and populate the memtable before running the test directly in the perf-query tool.

adding an option for perf-simple-query to stress tombstone heavy workloads (row, cell, partition tombstones) makes lots of sense, we can include it in QA process.

lkshminarayanan replied:
> > @raphaelsc I was finally able to modify `perf_simple_query` to test this PR as per your suggestion.
> > The `perf_simple_query` now
> > 
> > * sets gc_grace_seconds = 0s to the schema
> > * writes 10K partitions, with 33% tombstones
> > * waits for a second to let tombstones expire
> > * flushes all the tables and the commitlog
> > * drops the sstable caches
> > * writes 1K partitions in the memtable
> > * and then finally runs the read test.
> > 
> > This is the result of `build/Dev/scylla perf-simple-query --smp=1 --flush --duration=300` :
> > ```
> > // Before this Fix
> > // ---------------
> > 42081.34 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.1 tasks/op,   70067 insns/op,   32260 cycles/op,        0 errors)
> > 56577.90 tps ( 69.9 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   58029 insns/op,   24112 cycles/op,        0 errors)
> > 58074.16 tps ( 69.8 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57945 insns/op,   23777 cycles/op,        0 errors)
> > 58128.32 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57887 insns/op,   23775 cycles/op,        0 errors)
> > 58102.11 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57919 insns/op,   23766 cycles/op,        0 errors)
> > 
> > ...
> > ...
> >          throughput: mean=58113.29 standard-deviation=948.09 median=58207.25 median-absolute-deviation=118.37 maximum=58460.52 minimum=42081.34
> > instructions_per_op: mean=57968.85 standard-deviation=701.93 median=57927.94 median-absolute-deviation=44.95 maximum=70066.91 minimum=57815.54
> >   cpu_cycles_per_op: mean=23770.54 standard-deviation=494.48 median=23738.03 median-absolute-deviation=42.94 maximum=32259.76 minimum=23646.28
> > 
> > 
> > // After this Fix
> > // --------------
> > 41921.32 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   70069 insns/op,   32395 cycles/op,        0 errors)
> > 57360.31 tps ( 69.8 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   58091 insns/op,   24066 cycles/op,        0 errors)
> > 58126.02 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57907 insns/op,   23779 cycles/op,        0 errors)
> > 58348.81 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57924 insns/op,   23676 cycles/op,        0 errors)
> > 58266.16 tps ( 69.7 allocs/op,   0.0 logallocs/op,  14.3 tasks/op,   57898 insns/op,   23726 cycles/op,        0 errors)
> > ...
> > ...
> >          throughput: mean=58209.50 standard-deviation=964.09 median=58292.08 median-absolute-deviation=123.26 maximum=58661.90 minimum=41921.32
> > instructions_per_op: mean=57977.62 standard-deviation=701.51 median=57936.37 median-absolute-deviation=43.33 maximum=70068.51 minimum=57839.51
> >   cpu_cycles_per_op: mean=23739.14 standard-deviation=505.08 median=23709.30 median-absolute-deviation=42.39 maximum=32395.37 minimum=23573.21
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > WDYT? Do we have to track max sstable timestamp for each group reduce to short circuit the memtable lookup ?
> 
> just to make sure, we now can see the calls to new function to take into account memtable timestamp, right?

@raphaelsc yes I can see them now.

avikivity replied:
The results are too volatile. The first second results show 70k insn/op, the later results show 57k insn/op.

avikivity replied:
And yes, we probably need a short-circuit somewhere. It's rare that the tombstone in sstables is newer than data in memtables, so a short-circuit test would usually succeed.

lkshminarayanan replied:
> The results are too volatile. The first second results show 70k insn/op, the later results show 57k insn/op.

@avikivity That was due to a mistake on my part. Initially, I ran the performance tool without setting the CPU frequency, so it ran at the maximum CPU frequency, resulting in significant variance within the test as well. The later results were obtained after running the test at the lowest CPU frequency to ensure a consistent comparison between the before and after statistics.

avikivity replied:
insn/op and allocs/op wouldn't change due to clock frequency

raphaelsc replied:
maybe we're populating the cache on first iteration and hitting the cache mostly in later iterations. we need to drop the cache before each iteration for fairness. I am not sure the latter results are even valid, if we're not reading from the sstables.

lkshminarayanan replied:
@raphaelsc You are right. I have updated the perf tool to invalidate cache before each iteration. I have also added the short circuit based on sstable max timestamp. This is the perf results after these changes :

`build/Dev/scylla perf-simple-query --smp=1 --flush --duration=300` :

```
// Before this Fix
// ---------------
41409.53 tps ( 81.3 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69903 insns/op,   32786 cycles/op,        0 errors)
41460.87 tps ( 81.3 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69921 insns/op,   32742 cycles/op,        0 errors)
41261.20 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69915 insns/op,   32908 cycles/op,        0 errors)
41295.05 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69980 insns/op,   32845 cycles/op,        0 errors)
41598.98 tps ( 81.3 allocs/op,   2.6 logallocs/op,  18.1 tasks/op,   69827 insns/op,   32663 cycles/op,        0 errors)
...
...

         throughput: mean=41246.72 standard-deviation=197.71 median=41270.69 median-absolute-deviation=91.38 maximum=41617.56 minimum=40356.48
instructions_per_op: mean=69958.46 standard-deviation=83.16 median=69952.85 median-absolute-deviation=50.02 maximum=70273.18 minimum=69781.49
  cpu_cycles_per_op: mean=32897.40 standard-deviation=108.57 median=32894.00 median-absolute-deviation=67.25 maximum=33343.41 minimum=32647.84

// After this Fix
// --------------
41496.64 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69909 insns/op,   32665 cycles/op,        0 errors)
41603.10 tps ( 81.5 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69977 insns/op,   32639 cycles/op,        0 errors)
41507.06 tps ( 81.5 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   70000 insns/op,   32684 cycles/op,        0 errors)
41406.51 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69929 insns/op,   32760 cycles/op,        0 errors)
41521.08 tps ( 81.4 allocs/op,   2.6 logallocs/op,  18.2 tasks/op,   69901 insns/op,   32663 cycles/op,        0 errors)
...
...

         throughput: mean=41463.94 standard-deviation=176.43 median=41496.70 median-absolute-deviation=85.96 maximum=41791.34 minimum=40672.28
instructions_per_op: mean=69932.75 standard-deviation=76.33 median=69924.32 median-absolute-deviation=45.06 maximum=70263.93 minimum=69722.66
  cpu_cycles_per_op: mean=32709.25 standard-deviation=96.81 median=32700.68 median-absolute-deviation=61.39 maximum=33220.06 minimum=32496.67
```

lkshminarayanan commented:
Changelog V2 :
 - rebased on top of latest
 - Updated code to search only the storage group that has the token (Ref : [comment](https://github.com/scylladb/scylladb/pull/20985#discussion_r1790202731))

denesb commented on replica/table.cc:
```diff
@@ -2523,6 +2523,25 @@ table::make_partition_presence_checker(lw_shared_ptr<const sstables::sstable_set
     };
 }
 
+max_purgeable_fn table::get_max_purgeable_fn_for_compacting_reader() const {
```
This is too generic, a better name would be `get_max_purgeable_fn_for_cache_underlying_reader()`.
This name specifies that this is specific to the cache's underlying reader, which already tells you why sstables need not be considered (because all sstables that contain the read range are already included in the read).

lkshminarayanan replied:
Done.

lkshminarayanan commented:
Changelog V3 : renamed `get_max_purgeable_fn_for_compacting_reader` to `get_max_purgeable_fn_for_cache_underlying_reader`

raphaelsc commented on replica/table.cc:
```diff
@@ -2517,6 +2523,25 @@ table::make_partition_presence_checker(lw_shared_ptr<const sstables::sstable_set
+            // see get_max_purgeable_timestamp() in compaction.cc for comments on choosing min timestamp
+            api::timestamp_type memtable_min_timestamp = is_shadowable ? mt.get_min_live_row_marker_timestamp() : mt.get_min_live_timestamp();
+            if (memtable_min_timestamp <= max_purgeable_timestamp && mt.contains_partition(dk)) {
+                max_purgeable_timestamp = memtable_min_timestamp;
```
not that it matters much in practice, but there's an off-by-one here, we can bypass contains_partition() check when min_timestamp == max_purgeable, since it won't have any effect on the current max_purgeable.

lkshminarayanan replied:
Ack. Fixed it in v4.

lkshminarayanan commented:
Changelog V4 : Fixes this https://github.com/scylladb/scylladb/pull/20985#discussion_r1792374138

avikivity commented:
Could we not move the compacting reader behind the memtable+sstable merge?

raphaelsc commented:
> Could we not move the compacting reader behind the memtable+sstable merge?

I thought about that, but we cannot do it because the compacting reader happens in the cache data source, to avoid populating it with garbage. Then the cache reader has later its data combined with memtable.

avikivity commented:
Please add the commit that introduced the regression to the cover letter.

raphaelsc commented on replica/table.cc:
```diff
@@ -530,6 +530,16 @@ void compaction_group::backlog_tracker_adjust_charges(const std::vector<sstables
+    });
+    _maintenance_sstables->for_each_sstable([this] (const sstables::shared_sstable& sst) {
+        _max_sstable_timestamp = std::max(_max_sstable_timestamp, sst->get_stats_metadata().max_timestamp);
+    });
```
we can pass const sstables::sstable_set& here, and only iterate through set which had changes. I initially thought it was enough to track max sstable timestamp (seen), since it's monotonically increasing, but it's better to not rely on assumption that newer sstables will only be added through do_add_sstable(). it might change later.

raphaelsc replied:
ah I understand you iterate through both sets, so it can track the actual max timestamp. not the max timestamp seen, but I guess the latter is fine as long as it's documented?

raphaelsc replied:
maybe we can extend sstable_set to provide max_timestamp (currently it provides bytes on disk stat, see: https://github.com/scylladb/scylladb/commit/f54ab4827334373dfa203a0285fb090d212654b5.

```
sstables_set_stats {
    uint64_t bytes_on_disk;
    api::timestamp max_timestamp;
    
    sstables_set_stats& operator+=(const sstables_set_stats& other) noexcept;
}

then you change 

sstable_set_impl:
uint64_t add_bytes_on_disk(uint64_t delta) noexcept;

to become
sstable_set_stats update_stats(const shared_sstable&) noexcept;

sstable_set_impl:
virtual uint64_t bytes_on_disk() const noexcept

becomes
virtual sstable_set_stats stats() const noexcept
```

lkshminarayanan replied:
I have updated the compaction_group to track the maximum seen timestamp as discussed.

lkshminarayanan commented:
Changelog V5 : 
 - Track maximum seen timestamp rather than actual maximum in compaction_group

raphaelsc commented on replica/table.cc:
```diff
@@ -2531,6 +2531,32 @@ table::make_partition_presence_checker(lw_shared_ptr<const sstables::sstable_set
+        sg.for_each_compaction_group([&dk, is_shadowable, &max_purgeable_timestamp] (const compaction_group_ptr& cg) {
+            const auto& mt = cg->memtables()->active_memtable();
+            // see get_max_purgeable_timestamp() in compaction.cc for comments on choosing min timestamp
+            api::timestamp_type memtable_min_timestamp = is_shadowable ? mt.get_min_live_row_marker_timestamp() : mt.get_min_live_timestamp();
```
we can later extract it into a function for avoiding knowledge duplication.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12511/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12511/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: topology_custom/test_compacting_reader_tombstone_gc
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12511/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1225/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/12313/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/4291/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/12511/)
#### Build Details:
- Duration: 7 hr 30 min
- Builder: i-0e7cb63d3538685f0 (m5ad.8xlarge)

denesb commented:
> `build/Dev/scylla perf-simple-query --smp=1 --flush` :

In the future, please use `release` for performance measurements.
